content/learn/beginner/inputs-1.md

```diff
@@ -0,0 +1,169 @@
+number of required arguments, the remaining arguments are considered optional. You cannot define optional positional
+arguments using `parseArgs`.
+
+After that, we can provide a custom error message that will be displayed if the arguments are not valid. Passing an
+empty string `""` will use a generated one based on the defined arguments. This is useful for providing more context to
+the user about what went wrong.
+
+The `...carg` is a variadic argument, that is, it can take any number of arguments.
+Each `carg` is a single argument definition. The `carg` function has the following syntax:
+
+```go
+{{ carg <"type"> <"description"> }}
+```
+
+Following types are supported:
+
+- `int` (whole number)
+- `float` (decimal number)
+- `string` (text)
+- `user` (user mentions, resolves to the [user](https://docs.yagpdb.xyz/reference/templates#user) structure)
+- `userid` (mentions or user IDs, resolves to the ID itself)
+- `member` (mentions or user IDs, resolves to the [member](https://docs.yagpdb.xyz/reference/templates#member)
+ structure)
+- `channel` (channel mention or ID, resolves to the channel structure)
+- `role` (role name or ID, resolves as type _\*discordgo.Role_)
+- `duration` (duration that is human-readable, i.e `10h5m` or `10 hour 5 minutes` would both resolve to the same
+   duration)
+
+The `description` is a human-readable description of the argument. This is used in the error message if the argument is
+not valid.
+
+Combining all of this, let's create a custom command that takes two arguments: a coolness level and a user to apply it
+to.
+
+```go
+{{ $args := parseArgs 2 "" (carg "int" "coolness level") (carg "member" "target member") }}
+```
+
+Note that we are using the `member` type here --- in general, it is recommended to use the `member` type for
+user-related arguments, as it validates that the target user is also a member of the server. There are very limited
+use-cases for `user` and `userid`, which will be left as an exercise to the reader.
+
+### Accessing Arguments
+
+Currently, our code doesn't do anything with the arguments. To access the arguments, we use the `.Get` method on the
+`$args` variable. The syntax is as follows:
+
+```go
+{{ $args.Get <index> }}
+```
+
+The `index` is the index of the argument, starting from 0. The arguments are stored in the order they are defined in the
+`parseArgs` function call. Let us now modify our custom command to access these arguments:
+
+```go
+{{ $args := parseArgs 2 "" (carg "int" "coolness level") (carg "member" "target member") }}
+
+coolness: {{ $args.Get 0 }}
+member: {{ ($args.Get 1).Nick }}
+```
+
+### Validating Arguments
+
+#### Specifying Valid Ranges
+
+Now, we want to limit the coolness level to a number between 0 and 100. We can do this by adding a simple check:
+
+```go
+{{ $args := parseArgs 2 "" (carg "int" "coolness level") (carg "member" "target member") }}
+
+{{ if or (gt (toInt $args.Get 0) 100) (lt (toInt $args.Get 0) 0) }}
```

Comment:
this doesn't work correctly; `toInt $args.Get 0` means `toInt ($args.Get) 0`. Also, the guard clause obfuscates rather than helps here, IMO.

Reply:
The guard clause is indeed meant to obfuscate this code a little bit; I
want to show that it is just way easier to use `parseArg`'s validation
feature, which the following paragraph should hopefully illustrate.